File: leetcode/reverse-nodes-in-k-group/reverse-nodes-in-k-group.cpp
/**************************************
* @author dujiong
* @date 2016.12.1
* @version V0.1
**************************************/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *reverseKGroup(ListNode *head, int k) {
        if(head == NULL || k==1){
            return head;
        }
       	auto node = head;
        for(int i=0;i<k;++i){
            if(!node) return head;
            node = node->next;
        }
		//递归实现
        auto new_head = ReverseList(head, node);	
        head->next = reverseKGroup(node, k);
        return new_head;
    }
private:
    ListNode* ReverseList(ListNode* pHead, ListNode* pEnd) {	//调用ReverseList(pHead, pEnd)实现[pHead, pEnd->prev](不包括pEnd)的反转
        if(pHead->next == pEnd){
            return NULL;
        }
        ListNode* pNext = NULL;
        ListNode* pPrev = NULL;
       	while(pHead != pEnd){
            pNext = pHead->next;
            pHead->next = pPrev;
            pPrev = pHead;
            pHead = pNext;
        }
        return pPrev;
    }
};

File: leetcode-golang/Reverse Nodes in k-Group/Reverse Nodes in k-Group.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

func reverseKGroup(head *ListNode, k int) *ListNode {
	if head == nil || head <= 1 {
		return head
	}
	cur := head
	nodesLen := 0
	for cur != nil {
		nodesLen++
		cur = cur.Next
	}
	dummp := &ListNode{
		Next: head,
	}
	pre := dummy
	for ; nodesLen >= k; nodesLen -= k {
		cur := pre.Next
		for i := 1; i < k; i++ {
			temp := cur.Next
			cur.Next = temp.Next
			temp.Next = pre.Next
			pre.Next = temp
		}
		pre = cur
		cur = cur.Next
	}
	return dummy.Next
}

func reverseKGroup(head *ListNode, k int) *ListNode {
	cur := head
	for i := 0; i < k; i++ {
		if cur == nil {
			return head
		}
		cur = cur.Next
	}
	newHead := reverse(head, cur)
	head.Next = reverseKGroup(cur, k) //cur保存的是下一段链表的开始节点
	return newHead
}

func reverse(head *ListNode, tail *ListNode) *ListNode {
	pre := new(ListNode) //pre := tail
	for head != tail {
		temp := head.Next
		head.Next = pre
		pre = head
		head = temp
	}
	return pre
}

// func reverseList(head *ListNode) *ListNode {
//     if head == nil || head.Next == nil {
//         return head
//     }
//     var prev *ListNode
//     for head != nil {
//         n := head.Next
//         head.Next = prev
//         prev = head
//         head = n
//     }
//     return prev
// }

//非递归，较难理解
//以1->2->3->4->5，k=3为例进行分析
func reverseKGroup(head *ListNode, k int) *ListNode {
	if head == nil || k <= 1 {
		return head
	}
	cur := head
	nodesLen := 0
	for cur != nil {
		nodesLen++
		cur = cur.Next
	}
	dummy := &ListNode{
		Next: head,
	}
	pre := dummy
	for ; nodesLen >= k; nodesLen -= k {
		cur := pre.Next
		for i := 1; i < k; i++ {
			temp := cur.Next
			cur.Next = temp.Next
			temp.Next = pre.Next
			pre.Next = temp
		}
		pre = cur
		cur = cur.Next
	}
	return dummy.Next
}